pageserver/src/walreceiver.rs

```diff
@@ -1,170 +1,567 @@
+    >,
+) -> ControlFlow<anyhow::Result<()>, ()> {
+    match timeline_updates_receiver.recv().await {
+        Some(update) => {
+            info!("Processing timeline update: {update:?}");
+            match update {
+                LocalTimelineUpdate::Remove(id) => {
+                    match local_timeline_wal_receivers.get_mut(&id.tenant_id) {
+                        Some(timelines) => {
+                            if let Some(handles) = timelines.remove(&id.timeline_id) {
+                                let wal_receiver_shutdown_result = handles.shutdown(id).await;
+                                if wal_receiver_shutdown_result.is_err() {
+                                    return ControlFlow::Break(wal_receiver_shutdown_result)
+                                }
+                            };
+                            if timelines.is_empty() {
+                                let state_change_result = change_tenant_state(id.tenant_id, TenantState::Idle).await;
+                                if state_change_result.is_err() {
+                                    return ControlFlow::Break(state_change_result)
+                                }
+                            }
+                        }
+                        None => warn!("Timeline {id} does not have a tenant entry in wal receiver main thread"),
+                    };
+                }
+                LocalTimelineUpdate::Add(new_id, timeline) => {
+                    let timelines = local_timeline_wal_receivers
+                        .entry(new_id.tenant_id)
+                        .or_default();
+
+                    if timelines.is_empty() {
+                        let state_change_result =
+                            change_tenant_state(new_id.tenant_id, TenantState::Active).await;
+                        if state_change_result.is_err() {
+                            return ControlFlow::Break(state_change_result);
+                        }
+                    } else if let Some(old_entry_handles) = timelines.remove(&new_id.timeline_id) {
+                        warn!(
+                            "Readding to an existing timeline {new_id}, shutting the old wal receiver down"
+                        );
+                        let wal_receiver_shutdown_result = old_entry_handles.shutdown(new_id).await;
+                        if wal_receiver_shutdown_result.is_err() {
+                            return ControlFlow::Break(wal_receiver_shutdown_result);
+                        }
+                    }
+
+                    let (connect_timeout, max_retries) =
+                        match fetch_tenant_settings(new_id.tenant_id).await {
+                            Ok(settings) => settings,
+                            Err(e) => return ControlFlow::Break(Err(e)),
+                        };
+
+                    let mut connection_failed = true;
+                    for attempt in 0..max_retries.get() {
```

Comment:
I don't think we need any tracking of failures here, on task start. Broker task can fail later at any moment (you'd just see watch channel closed at current coding), so that wouldn't help anyway. And currently no one ever restarts broker tasks, that's wrong. We need infinite reconnect-to-etcd logic either inside the task or in walreceiver_main_thread level. At some point I realized that sharing etcd client among tasks is good for performance (still need to check to be absolutely sure, but looks like they share underlying http/2 connection and request multiplexing), so the latter is preferrable.

However, problem with etcd shouldn't immediately kill the current walreceiver connection I guess, so it might be better to track walreceiver state in shared memory so that next broker task can pick it up.

Reply:
Replaced with the infinite loop, no more killing on that thing is done.

I've not placed this particular state into some shared mem, since turned out that the state we need for the broker is easier to recreate, but could be improved later if needed.

We would want to improve the subscription management based on the recent discussions, so we keep a single conneciton for multiple subscriptions, but that requires more work on our `broker` crate and not done right now.